tests/test_add_new_method.py

```diff
@@ -0,0 +1,879 @@
+"""
+Test suite for dynamically adding new contraceptive methods.
+
+This suite validates the new feature that allows adding contraceptive methods
+during a simulation via interventions. It tests:
+- Basic method addition
+- Switching matrix expansion
+- Integration with different contraception modules
+- Edge cases and error handling
+"""
+
+import numpy as np
+import pytest
+import sciris as sc
+import fpsim as fp
+from fpsim.methods import NewMethodConfig
+
+
+def make_sim(**kwargs):
+    """Create a small test simulation."""
+    default_kwargs = dict(
+        n_agents=500,
+        location='kenya',
+        start_year=2000,
+        end_year=2015,
+    )
+    default_kwargs.update(kwargs)
+    return fp.Sim(**default_kwargs)
+
+
+# ==============================================================================
+# Test 1: Basic method addition
+# ==============================================================================
+def test_basic_method_addition():
+    """Test that a new method can be added and appears in the simulation."""
+    sc.heading('Testing basic method addition...')
+    
+    # Create a new method
+    new_method = fp.Method(
+        name='test_method',
+        label='Test Method',
+        efficacy=0.95,
+        modern=True,
+        dur_use=fp.methods.ln(6, 2)
+    )
+    
+    # Create intervention
+    mod = fp.MethodIntervention(year=2010, label='Add Test Method')
+    mod.add_method(
+        method=new_method,
+        copy_from_row='inj',
+        copy_from_col='inj',
+        initial_share=0.3
+    )
+    
+    intv = mod.build()
+    sim = make_sim(interventions=intv)
+    sim.run()
+    
+    # Verify method was added
+    cm = sim.connectors.contraception
+    assert 'test_method' in cm.methods, "New method not found in contraception module"
+    assert cm.methods['test_method'].idx == 10, "New method has wrong index"
+    
+    # Verify method_mix was resized
+    fp_mod = sim.connectors['fp']
+    assert fp_mod.method_mix.shape[0] == 11, f"method_mix not resized correctly: {fp_mod.method_mix.shape}"
+    
+    # Verify method has some adoption
+    final_adoption = fp_mod.method_mix[10, -1] * 100
+    assert final_adoption > 0, "New method has zero adoption"
+    
+    print(f'✓ Basic method addition successful (adoption: {final_adoption:.2f}%)')
+    return sim
+
+
+# ==============================================================================
+# Test 2: Switching matrix expansion
+# ==============================================================================
+def test_switching_matrix_expansion():
+    """Test that switching matrix is properly expanded."""
+    sc.heading('Testing switching matrix expansion...')
+    
+    new_method = fp.Method(
+        name='new_inj',
+        label='New Injectable',
+        efficacy=0.98,
+        modern=True,
+        dur_use=fp.methods.ln(8, 2)
+    )
+    
+    mod = fp.MethodIntervention(year=2010)
+    mod.add_method(
```

Comment:
I'm curious whether this also supports a workflow where people want to manually upload their own switching matrix